### ISSUE TITLE
Strange behaviour on tables with NULL values

### ISSUE THREAD

**Issue by wnkz** [crate/crate#2188]
I noticed a strange table behaviour when performing `DELETE` on tables with `NULL` values when inserted with a query.
I'm testing with Crate 0.49.3 on Docker ; here are the steps to reproduce:

```sql
create table t1 (id integer, label string) ;
insert into t1 (id, label) values (0, 'zero'), (1, 'one'), (2, 'two') ;
insert into t1 (id) values (3), (4), (5) ;
select * from t1 order by id ;
select * from t1 where label is not null order by id ;
select * from t1 where label is null order by id ;
create table t2 (id integer, label string) ;
insert into t2 (id, label) (select id, label from t1) ;
select * from t2 order by id ;
select * from t2 where label is not null order by id ;
select * from t2 where label is null order by id ;
delete from t2 where label is null ;
select * from t2 order by id ;
select * from t2 where label is not null order by id ;
select * from t2 where label is null order by id ;
```

And here is the output:

```sql
cr> create table t1 (id integer, label string) ;
CREATE OK (0.528 sec)
cr> insert into t1 (id, label) values (0, 'zero'), (1, 'one'), (2, 'two') ;
INSERT OK, 3 rows affected (0.097 sec)
cr> insert into t1 (id) values (3), (4), (5) ;
INSERT OK, 3 rows affected (0.018 sec)
cr> select * from t1 order by id ;
+----+-------+
| id | label |
+----+-------+
|  0 | zero  |
|  1 | one   |
|  2 | two   |
|  3 | NULL  |
|  4 | NULL  |
|  5 | NULL  |
+----+-------+
SELECT 6 rows in set (0.120 sec)
cr> select * from t1 where label is not null order by id ;
+----+-------+
| id | label |
+----+-------+
|  0 | zero  |
|  1 | one   |
|  2 | two   |
+----+-------+
SELECT 3 rows in set (0.033 sec)
cr> select * from t1 where label is null order by id ;
+----+-------+
| id | label |
+----+-------+
|  3 |  NULL |
|  4 |  NULL |
|  5 |  NULL |
+----+-------+
SELECT 3 rows in set (0.009 sec)
cr> create table t2 (id integer, label string) ;
CREATE OK (0.072 sec)
cr> insert into t2 (id, label) (select id, label from t1) ;
INSERT OK, 6 rows affected (0.034 sec)
cr> select * from t2 order by id ;
+----+-------+
| id | label |
+----+-------+
|  0 | zero  |
|  1 | one   |
|  2 | two   |
|  3 | NULL  |
|  4 | NULL  |
|  5 | NULL  |
+----+-------+
SELECT 6 rows in set (0.013 sec)
cr> select * from t2 where label is not null order by id ;
+----+-------+
| id | label |
+----+-------+
|  0 | zero  |
|  1 | one   |
|  2 | two   |
+----+-------+
SELECT 3 rows in set (0.014 sec)
cr> select * from t2 where label is null order by id ;
+----+-------+
| id | label |
+----+-------+
|  3 |  NULL |
|  4 |  NULL |
|  5 |  NULL |
+----+-------+
SELECT 3 rows in set (0.009 sec)
cr> delete from t2 where label is null ;
DELETE OK, -1 rows affected (0.062 sec)
cr> select * from t2 order by id ;
+----+-------+
| id | label |
+----+-------+
|  0 | zero  |
|  1 | one   |
|  2 | two   |
+----+-------+
SELECT 3 rows in set (0.007 sec)
cr> select * from t2 where label is not null order by id ;
+----+-------+
| id | label |
+----+-------+
|  0 | zero  |
|  1 | one   |
|  2 | two   |
+----+-------+
SELECT 3 rows in set (0.009 sec)
cr> select * from t2 where label is null order by id ;
+----+-------+
| id | label |
+----+-------+
|  5 |  NULL |
+----+-------+
SELECT 1 row in set (0.007 sec)
```

Everything looks ok until the last query: `select * from t2 where label is null order by id ;` that should return an empty set but instead shows a row that should have been deleted **and** which doesn't show in the `select * from t2 ;`.

FYI, nothing special shows on the server logs:

```
[...]
[2015-06-30 10:16:16,608][INFO ][cluster.service          ] [Aunt May Parker] new_master [Aunt May Parker][OCRcLJGaR4av4bAru4hMPQ][7a4b59a0a772][inet[/172.17.0.2:4300]]{http_address=http://172.17.0.2:4200}, reason: zen-disco-join (elected_as_master)
[2015-06-30 10:16:16,618][INFO ][node                     ] [Aunt May Parker] started
[2015-06-30 10:16:16,638][INFO ][gateway                  ] [Aunt May Parker] recovered [0] indices into cluster_state
[2015-06-30 10:19:20,285][INFO ][cluster.metadata         ] [Aunt May Parker] [t1] creating index, cause [api], shards [5]/[1], mappings [default]
[2015-06-30 10:20:07,121][INFO ][cluster.metadata         ] [Aunt May Parker] [t2] creating index, cause [api], shards [5]/[1], mappings [default]
```

Those are the simplest steps to reproduce I could find but this happens on more complex tables and is driving us crazy. Are we doing anything wrong and/or is there any workaround ?

**Comment by chaudum** [comment]
hi @wnkz 
we found the problem and there's already a fix in master and `0.49` but not released yet. will let you know when it's out.

**Comment by wnkz** [comment]
Thanks @chaudum ; is this supposed to be fixed in 0.49.4 or the next one ?

**Comment by chaudum** [comment]
@wnkz it's released now (`0.49.5`) https://github.com/crate/crate/blob/0.49/CHANGES.txt#L25
